File: 2015/Day_12/Perl/part_2.pl
#!/usr/bin/perl

use JSON::PP;
use Scalar::Util qw(looks_like_number);
use v5.20;
use warnings;

sub json_sums{
    my $summation = 0;
    my $json_item = $_[0];
    if (ref($json_item) eq "ARRAY"){
        for my $item (@{$json_item}){
            $summation += json_sums($item);
        }
    }
    elsif (looks_like_number($json_item)){
        return $json_item;
    }
    elsif (ref($json_item) eq "HASH"){
        my @values = (values %{$json_item});
        for my $value (@values){
            if ($value eq "red"){
                return 0;
            }
        }
        for my $value (@values){
            if (looks_like_number($value)){
                $summation += $value;
            }
            elsif (ref($value) eq "ARRAY"){
                $summation += json_sums($value);
            }
            elsif (ref($value) eq "HASH"){
                $summation += json_sums($value);
            }
            else{
                $summation += 0;
            }
        }
    }
    return $summation;
}

open(ELFJSON, "../input.txt") || die "NO ELF JSON!";

my $elf_json = decode_json <ELFJSON>;

# test
# say "Array!" if ref($elf_json) eq 'ARRAY';
# say "Hash!" if ref($elf_json) eq 'HASH';

my $final_sum = json_sums($elf_json);

say "The total is $final_sum";

File: 2015/Day_12/Ruby/part_2.rb
require 'json'
require "../../input_parsing/parse_input"


def json_sums(json_item)
    summation = 0
    if json_item.is_a?(Array)
        json_item.each do |item|
            summation += json_sums(item)
        end
    elsif json_item.is_a?(Integer)
        return json_item
    elsif json_item.is_a?(String)
        return 0
    else
        json_item.each_pair do |key, value|
            if json_item.values.include? 'red'
                return 0
            elsif value.is_a?(Integer)
                summation += value
            elsif value.is_a?(Array)
                summation += json_sums(value)
            elsif value.is_a?(Hash)
                summation += json_sums(value)
            else
                summation += 0
            end
        end
    end
    return summation
end

if $PROGRAM_NAME == __FILE__
    elf_json = input_per_line('../input.txt')
    parsed_elf_json = JSON.parse(elf_json[0])
    total = json_sums(parsed_elf_json)
    puts "The sum is #{total}"
end
